Regression(247804@main): Network process is terminated while fetching website data

szewai opened the pull request:
#### 5867c9a72edd6825d5ace624f462c9161f8ba673
<pre>
Regression(247804@main): Network process is terminated while fetching website data
<a href="https://bugs.webkit.org/show_bug.cgi?id=241830">https://bugs.webkit.org/show_bug.cgi?id=241830</a>

Reviewed by Chris Dumez.

229885@main moves the ownership of NetworkProcessProxy from WebProcessPool to WebsiteDataStore, and we started to have
an issue that network process never exits, because we have default WebsiteDataStore that is never destroyed.
236145@main tries to make network process exit by setting NetworkProcessProxy::defaultNetworkProcess() to null when last
WebProcessPool is gone. However, it didn&apos;t achieve the goal, because it failed to clear the strong reference in
WebsiteDataStore, which means network process can stay as long as default WebsiteDataStore.
247804@main fixes the issue above by clearing the reference in WebsiteDataStore and now network process will actually
be terminated when last WebProcessPool is destroyed. However, network process can be terminated when WebsiteDataStore is
using it (e.g. waiting for task reply as in new test).

We may fix this issue by protecting NetworkProcessProxy when WebsiteDataStore is using network process, but this is
non-intuitive and with this approach we may accidentally introduce the bug again (i.e. we fail to protect
NetworkProcessProxy in some case and network process gets terminated while in use). Also, network process can carry
some session states in memory, so it feels weird that the states are lost while WebsiteDataStore is still alive. The
correct solution may be let default WebsiteDataStore be destroyed when it&apos;s not in use.

Therefore, to fix existing bug, the patch restores the old behavior: network process is terminated when last
WebsiteDataStore using it is destroyed.

API test: NetworkProcess.TerminateWhenNoWebsiteDataStore
          NetworkProcess.TerminateWhenNoDefaultWebsiteDataStore
          WKWebsiteDataStore.FetchPersistentWebStorage

* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::defaultNetworkProcess):
(WebKit::NetworkProcessProxy::ensureDefaultNetworkProcess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::~WebProcessPool):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::WebsiteDataStore):
(WebKit::WebsiteDataStore::~WebsiteDataStore):
(WebKit::protectedDefaultDataStore):
(WebKit::globalDefaultDataStore):
(WebKit::WebsiteDataStore::defaultDataStore):
(WebKit::WebsiteDataStore::deleteDefaultDataStoreForTesting):
(WebKit::WebsiteDataStore::removeNetworkProcessReference): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::networkProcessIfExists):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NetworkProcess.mm:
(TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/251983@main">https://commits.webkit.org/251983@main</a>
</pre>

cdumez reviewed (changes requested):
You needed to modify NetworkProcess.TerminateWhenUnused API which is evidence that you regressed things.

You managed to keep the test "mostly" passing by calling [WKWebsiteDataStore _deleteDefaultDataStoreForTesting], which is something clients cannot do.

So the result with this patch is the same as with the last patch, the network process doesn't exit if the default data store was ever used, which is not OK.

cdumez commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -132,17 +132,21 @@ Vector<Ref<NetworkProcessProxy>> NetworkProcessProxy::allNetworkProcesses()
-    if (!defaultNetworkProcess())
-        defaultNetworkProcess() = NetworkProcessProxy::create();
-    return *defaultNetworkProcess();
+    auto& networkProcesss = defaultNetworkProcess();
```
Typo: networkProcesss

cdumez reviewed (approved):
LGTM with typo fix.

szewai committed:
Regression(247804@main): Network process is terminated while fetching website data
https://bugs.webkit.org/show_bug.cgi?id=241830

Reviewed by Chris Dumez.

229885@main moves the ownership of NetworkProcessProxy from WebProcessPool to WebsiteDataStore, and we started to have
an issue that network process never exits, because we have default WebsiteDataStore that is never destroyed.
236145@main tries to make network process exit by setting NetworkProcessProxy::defaultNetworkProcess() to null when last
WebProcessPool is gone. However, it didn't achieve the goal, because it failed to clear the strong reference in
WebsiteDataStore, which means network process can stay as long as default WebsiteDataStore.
247804@main fixes the issue above by clearing the reference in WebsiteDataStore and now network process will actually
be terminated when last WebProcessPool is destroyed. However, network process can be terminated when WebsiteDataStore is
using it (e.g. waiting for task reply as in new test).

We may fix this issue by protecting NetworkProcessProxy when WebsiteDataStore is using network process, but this is
non-intuitive and with this approach we may accidentally introduce the bug again (i.e. we fail to protect
NetworkProcessProxy in some case and network process gets terminated while in use). Also, network process can carry
some session states in memory, so it feels weird that the states are lost while WebsiteDataStore is still alive. The
correct solution may be let default WebsiteDataStore be destroyed when it's not in use.

Therefore, to fix existing bug, the patch restores the old behavior: network process is terminated when last
WebsiteDataStore using it is destroyed.

API test: NetworkProcess.TerminateWhenNoWebsiteDataStore
          NetworkProcess.TerminateWhenNoDefaultWebsiteDataStore
          WKWebsiteDataStore.FetchPersistentWebStorage

* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::defaultNetworkProcess):
(WebKit::NetworkProcessProxy::ensureDefaultNetworkProcess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::~WebProcessPool):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::WebsiteDataStore):
(WebKit::WebsiteDataStore::~WebsiteDataStore):
(WebKit::protectedDefaultDataStore):
(WebKit::globalDefaultDataStore):
(WebKit::WebsiteDataStore::defaultDataStore):
(WebKit::WebsiteDataStore::deleteDefaultDataStoreForTesting):
(WebKit::WebsiteDataStore::removeNetworkProcessReference): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::networkProcessIfExists):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NetworkProcess.mm:
(TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/251983@main
